File: python/plugins/page_flows_plugin.py
import json
from collections.abc import Iterator
from pathlib import Path

from models.document import Document
from plugins.md2html_plugin import Md2HtmlPlugin
from utils import relativize_relative_resource, first_not_none, UserError

MODULE_DIR = Path(__file__).resolve().parent


class PageFlow(Iterator):

    def __init__(self, title, pages, previous_page=None, current_page=None, next_page=None):
        self.title = title
        self.pages = pages
        self.previous = previous_page
        self.current = current_page
        self.next = next_page
        # For a logic-less template like Mustache the following calculated fields will help a lot.
        self.has_navigation = self.previous is not None or self.next is not None
        self.not_empty = self.pages is not None and len(self.pages) > 0

    def __iter__(self):
        return self.pages.__iter__()

    def __next__(self):
        return self.pages.__next__()


def process_page_flow(page_flow, output_file):
    pages = []
    previous_page = None
    current_page = None
    next_page = None

    for page in page_flow["items"]:
        new_page = dict(page)
        if page["external"]:
            new_page["current"] = False
            pages.append(new_page)
        else:
            is_current = page["link"] == output_file
            new_page["link"] = relativize_relative_resource(page["link"], output_file)
            new_page["current"] = is_current
            pages.append(new_page)
            if current_page is None:
                if is_current:
                    current_page = new_page
                else:
                    previous_page = new_page
            elif next_page is None:
                next_page = new_page

    if current_page is None:
        previous_page = None

    return PageFlow(page_flow["title"], pages, previous_page, current_page, next_page)


def unify_page_flow(page_flow):
    return {"title": page_flow.get("title") or "",
            "groups": page_flow.get("groups") or [],
            "items": page_flow.get("items") or []
            }


def unify_data(data):
    new_data = {}
    for k, v in data.items():
        if type(v) == list:
            new_data[k] = unify_page_flow({"items": v})
        elif type(v) == dict:
            new_data[k] = unify_page_flow(v)
        else:
            raise Exception(f"Wrong value type: {type(v).__name__}")
    return new_data


class PageFlowsPlugin(Md2HtmlPlugin):

    def __init__(self):
        super().__init__()
        self.initialize_done = False
        self.raw_data: dict = {}
        self.page_flows: dict = {}
        with open(MODULE_DIR.joinpath('page_flows_schema.json'), 'r',
                  encoding="utf-8") as schema_file:
            self.data_schema = json.load(schema_file)

    def accept_data(self, data):
        self.validate_data_with_schema(data, self.data_schema)
        data = unify_data(data)
        if self.raw_data:
            self.add_to_start(data)
        else:
            self.raw_data = data

    def add_to_start(self, data):
        if data is None:
            return
        for k, v in data.items():
            if k in self.raw_data:
                self.raw_data[k]["items"] = v["items"] + self.raw_data[k]["items"]
            else:
                self.raw_data[k] = v

    def add_to_end(self, data):
        if data is None:
            return
        for k, v in data.items():
            if k in self.raw_data:
                self.raw_data[k]["items"].extend(v["items"])
            else:
                self.raw_data[k] = v

    def initialize_data(self):
        page_flows = {}
        for k, v in self.raw_data.items():
            page_flow_items = []
            new_page = {}
            is_first = True
            for item in v["items"]:
                # Allows other arbitrary fields. This fields then may be used in the templates.
                new_page = dict(item)
                new_page["external"] = first_not_none(item.get("external"), False)
                new_page["first"] = is_first
                new_page["last"] = False

                page_flow_items.append(new_page)
                is_first = False
            new_page["last"] = True
            new_v = dict(v)
            new_v["items"] = page_flow_items
            page_flows[k] = new_v
        self.page_flows = page_flows

    def initialize(self, extra_plugin_data):
        self.assure_initialize_once()
        if bool(extra_plugin_data):
            self.validate_data_with_schema(extra_plugin_data, self.data_schema)
            extra_plugin_data = unify_data(extra_plugin_data)
            self.add_to_end(extra_plugin_data)
        self.initialize_data()

    def is_blank(self) -> bool:
        return not bool(self.page_flows)

    def variables(self, doc: Document) -> dict:
        page_flows = {}
        page_flow_groups = {}
        for k, v in self.page_flows.items():
            page_flow = process_page_flow(v, doc.output_file)
            page_flows[k] = page_flow
            for group in v["groups"]:
                page_flow_group = page_flow_groups.setdefault(group, [])
                page_flow_group.append(page_flow)
        variables = page_flows
        for g, pfs in page_flow_groups.items():
            if g in variables:
                raise UserError(f"Variable duplication error in plugin '{type(self).__name__}': "
                                f"group name is '{g}'")
            else:
                variables[g] = pfs
        return variables

    def assure_initialize_once(self):
        if self.initialize_done:
            raise Exception("Trying to initialize again.")
        self.initialize_done = True

File: java/src/main/java/world/md2html/plugins/PageFlowsPlugin.java
package world.md2html.plugins;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.TextNode;
import com.networknt.schema.JsonSchema;
import lombok.Getter;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.Document;
import world.md2html.utils.CheckedIllegalArgumentException;
import world.md2html.utils.JsonUtils;
import world.md2html.utils.UserError;
import world.md2html.utils.Utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static world.md2html.utils.JsonUtils.NODE_FACTORY;
import static world.md2html.utils.JsonUtils.loadJsonSchemaFromResource;
import static world.md2html.utils.Utils.relativizeRelativeResource;

public class PageFlowsPlugin extends AbstractMd2HtmlPlugin {

    private final JsonSchema pluginDataSchema =
            loadJsonSchemaFromResource("plugins/page_flows_schema.json");

    private Map<String, PageFlowRaw> data = null;
    private ObjectNode jsonData = null;

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {
        validateInputDataAgainstSchema(data, pluginDataSchema);
        data = unifyData(data);
        if (this.jsonData == null) {
            this.jsonData = (ObjectNode) data;
        } else {
            addToStart((ObjectNode) data);
        }
    }

    private JsonNode unifyData(JsonNode data) {
        ObjectNode newData = new ObjectNode(NODE_FACTORY);
        Iterator<Map.Entry<String, JsonNode>> fieldIterator = data.fields();
        while (fieldIterator.hasNext()) {
            Map.Entry<String, JsonNode> fieldEntry = fieldIterator.next();
            if (fieldEntry.getValue().isArray()) {
                ObjectNode pageFlow = new ObjectNode(NODE_FACTORY);
                pageFlow.set("items", fieldEntry.getValue());
                newData.set(fieldEntry.getKey(), unifyPageFlow(pageFlow));
            } else if (fieldEntry.getValue().isObject()) {
                newData.set(fieldEntry.getKey(), unifyPageFlow((ObjectNode) fieldEntry.getValue()));
            } else {
                throw new RuntimeException("Wrong value type: " + fieldEntry.getClass().getSimpleName());
            }
        }
        return newData;
    }

    private ObjectNode unifyPageFlow(ObjectNode pageFlow) {
        ObjectNode newPageFlow = new ObjectNode(NODE_FACTORY);
        newPageFlow.set("title", pageFlow.has("title") ? pageFlow.get("title") : new TextNode(""));
        newPageFlow.set("groups", pageFlow.has("groups") ? pageFlow.get("groups") : new ArrayNode(NODE_FACTORY));
        newPageFlow.set("items", pageFlow.has("items") ? pageFlow.get("items") : new ArrayNode(NODE_FACTORY));
        return newPageFlow;
    }

    private void addToStart(ObjectNode data) {
        if (data == null) {
            return;
        }
        for (Iterator<Map.Entry<String, JsonNode>> it = data.fields(); it.hasNext(); ) {
            Map.Entry<String, JsonNode> dataEntry = it.next();
            ObjectNode existingPageFlow = (ObjectNode) this.jsonData.get(dataEntry.getKey());
            if (existingPageFlow == null) {
                this.jsonData.set(dataEntry.getKey(), dataEntry.getValue());
            } else {
                ArrayNode existingPages = (ArrayNode) existingPageFlow.get("items");
                ArrayNode newPages = (ArrayNode) existingPageFlow
                        .set("items", new ArrayNode(NODE_FACTORY)).get("items");
                newPages.addAll((ArrayNode) dataEntry.getValue().get("items"));
                newPages.addAll(existingPages);
            }
        }
    }

    private void addToEnd(ObjectNode data) {
        if (data == null) {
            return;
        }
        for (Iterator<Map.Entry<String, JsonNode>> it = data.fields(); it.hasNext(); ) {
            Map.Entry<String, JsonNode> dataEntry = it.next();
            ObjectNode existingPageFlow = (ObjectNode) this.jsonData.get(dataEntry.getKey());
            if (existingPageFlow == null) {
                this.jsonData.set(dataEntry.getKey(), dataEntry.getValue());
            } else {
                ArrayNode existingPages = (ArrayNode) existingPageFlow.get("items");
                existingPages.addAll((ArrayNode) dataEntry.getValue().get("items"));
            }
        }
    }

    @Override
    public void initialize(JsonNode extraPluginData) throws ArgFileParseException {
        assureInitializeOnce();
        if (extraPluginData != null) {
            validateInputDataAgainstSchema(extraPluginData, pluginDataSchema);
            addToEnd((ObjectNode) unifyData(extraPluginData));
        }
        parsePluginData(this.jsonData);
    }

    private static class PageFlowRaw {
        @Getter private final String title;
        @Getter private final List<String> groups;
        @Getter private final List<Map<String, Object>> pages;

        public PageFlowRaw(String title, List<String> groups, List<Map<String, Object>> pages) {
            this.title = title;
            this.groups = groups;
            this.pages = pages;
        }
    }

    private void parsePluginData(JsonNode data) {
        Map<String, PageFlowRaw> pluginData = new LinkedHashMap<>();
        for (Iterator<Map.Entry<String, JsonNode>> it = data.fields(); it.hasNext(); ) {
            Map.Entry<String, JsonNode> pageFlowEntry = it.next();
            List<Map<String, Object>> pages = new ArrayList<>();
            Map<String, Object> page = new HashMap<>();
            boolean isFirst = true;
            JsonNode pageFlow = pageFlowEntry.getValue();
            for (Iterator<JsonNode> it1 = pageFlow.get("items").elements(); it1.hasNext(); ) {
                ObjectNode pageNode = (ObjectNode) it1.next();
                page = new HashMap<>();
                for (Iterator<Map.Entry<String, JsonNode>> it2 = pageNode.fields(); it2.hasNext(); ) {
                    Map.Entry<String, JsonNode> fieldEntry = it2.next();
                    page.put(fieldEntry.getKey(), JsonUtils.deJson(fieldEntry.getValue()));
                }
                enrichPage(page);
                page.put("first", isFirst);
                page.put("last", false);
                isFirst = false;
                pages.add(page);
            }
            page.put("last", true);
            List<String> groups = new ArrayList<>(pageFlow.get("groups").size());
            for (Iterator<JsonNode> it3 = pageFlow.get("groups").elements(); it3.hasNext(); ) {
                groups.add(it3.next().asText());
            }

            pluginData.put(pageFlowEntry.getKey(), new PageFlowRaw(pageFlow.get("title").asText(),
                    groups, pages));
        }
        this.data = pluginData;
    }

    @Override
    public boolean isBlank() {
        return this.data.isEmpty();
    }

    public Map<String, Object> variables(Document document) {
        Map<String, PageFlow> pageFlowVariables = new LinkedHashMap<>();
        Map<String, List<PageFlow>> pageFlowGroupVariables = new HashMap<>();
        this.data.forEach((k, v) -> {
            PageFlow pageFlow;
            try {
                pageFlow = processPageFlow(v, document.getOutput());
            } catch (CheckedIllegalArgumentException e) {
                throw new UserError("Error recalculating relative links '" + v +
                        "' of page flow '" + k + "' for page '" + document.getOutput() +
                        "': " + e.getMessage());
            }
            pageFlowVariables.put(k, pageFlow);
            v.getGroups().forEach(g -> {
                pageFlowGroupVariables.putIfAbsent(g, new ArrayList<>());
                List<PageFlow> groupPageFlows = pageFlowGroupVariables.get(g);
                groupPageFlows.add(pageFlow);
            });
        });
        pageFlowGroupVariables.keySet().forEach(key -> {
            if (pageFlowVariables.containsKey(key)) {
                throw new UserError("Variable duplication error in plugin '" +
                        this.getClass().getSimpleName() + "': group name is '" + key + "'");
            }
        });
        Map<String, Object> variables = new HashMap<>(pageFlowVariables);
        variables.putAll(pageFlowGroupVariables);
        return variables;
    }

    private static PageFlow processPageFlow(PageFlowRaw pageFlowRaw, String outputFile)
            throws CheckedIllegalArgumentException {

        List<Map<String, Object>> result = new ArrayList<>();
        Map<String, Object> previous = null;
        Map<String, Object> current = null;
        Map<String, Object> next = null;

        for (Map<String, Object> page : pageFlowRaw.getPages()) {
            Map<String, Object> newPage = new HashMap<>(page);
            if (Utils.isNullOrFalse(page.get("external"))) {
                result.add(newPage);
            } else {
                String link = (String) page.get("link");
                boolean isCurrent = link.equals(outputFile);
                newPage.put("link", relativizeRelativeResource(link, outputFile));
                newPage.put("current", isCurrent);
                result.add(newPage);
                if (current == null) {
                    if (isCurrent) {
                        current = newPage;
                    } else {
                        previous = newPage;
                    }
                } else if (next == null) {
                    next = newPage;
                }
            }
        }
        if (current == null) {
            previous = null;
        }
        return new PageFlow(pageFlowRaw.getTitle(), result, previous, current, next);
    }

    private static void enrichPage(Map<String, Object> page) {
        page.putIfAbsent("external", false);
        page.putIfAbsent("current", false);
    }

    private static class PageFlow implements Iterable<Map<String, Object>> {

        @Getter private final String title;
        @Getter private final List<Map<String, Object>> pages;
        @Getter private final Map<String, Object> previous;
        @Getter private final Map<String, Object> current;
        @Getter private final Map<String, Object> next;
        // For a logic-less template like Mustache the following calculated fields will help a lot.
        @Getter private final boolean has_navigation;
        @Getter private final boolean not_empty;

        private PageFlow(String title, List<Map<String, Object>> pages,
                Map<String, Object> previous,
                Map<String, Object> current, Map<String, Object> next) {
            this.title = title;
            this.pages = pages;
            this.previous = previous;
            this.current = current;
            this.next = next;
            this.has_navigation = this.previous != null || this.next != null;
            this.not_empty = this.pages != null && !this.pages.isEmpty();
        }

        @Override
        public Iterator<Map<String, Object>> iterator() {
            return pages.iterator();
        }
    }
}
